core/src/main/java/nl/nn/adapterframework/http/authentication/OAuthAccessTokenManager.java

```diff
@@ -0,0 +1,179 @@
+import com.nimbusds.oauth2.sdk.auth.ClientAuthentication;
+import com.nimbusds.oauth2.sdk.auth.ClientSecretBasic;
+import com.nimbusds.oauth2.sdk.auth.Secret;
+import com.nimbusds.oauth2.sdk.http.HTTPRequest;
+import com.nimbusds.oauth2.sdk.http.HTTPResponse;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.oauth2.sdk.token.AccessToken;
+
+import nl.nn.adapterframework.http.HttpSenderBase;
+import nl.nn.adapterframework.task.TimeoutGuard;
+import nl.nn.adapterframework.util.CredentialFactory;
+import nl.nn.adapterframework.util.Misc;
+import nl.nn.adapterframework.util.StreamUtil;
+
+public class OAuthAccessTokenManager {
+
+	private String tokenEndpoint;
+	private Scope scope;
+	private CredentialFactory client_cf;
+	private boolean useClientCredentialsGrant;
+	private HttpSenderBase httpSender;
+	
+	private AccessToken accessToken;
+	private long accessTokenRefreshTime;
+
+	public OAuthAccessTokenManager(String tokenEndpoint, String scope, CredentialFactory client_cf, boolean useClientCredentialsGrant, HttpSenderBase httpSender) {
+		this.tokenEndpoint = tokenEndpoint;
+		this.scope = StringUtils.isNotEmpty(scope) ? Scope.parse(scope) : null;
+		this.client_cf = client_cf;
+		this.useClientCredentialsGrant = useClientCredentialsGrant;
+		this.httpSender = httpSender;
+	}
+
+	public synchronized void retrieveAccessToken(Credentials credentials) throws HttpAuthenticationException {
+		AuthorizationGrant grant;
+
+		if (useClientCredentialsGrant) {
+			grant = new ClientCredentialsGrant();
+		} else {
+			String username = credentials.getUserPrincipal().getName();
+			Secret password = new Secret(credentials.getPassword());
+			grant = new ResourceOwnerPasswordCredentialsGrant(username, password);
+		}
+
+		// The credentials to authenticate the client at the token endpoint
+		ClientID clientID = new ClientID(client_cf.getUsername());
+		Secret clientSecret = new Secret(client_cf.getPassword());
+		ClientAuthentication clientAuth = new ClientSecretBasic(clientID, clientSecret);
+
+		try {
+			URI _tokenEndpoint = new URI(tokenEndpoint);
+			TokenRequest request = new TokenRequest(_tokenEndpoint, clientAuth, grant, scope);
+
+			try {
+				HTTPRequest httpRequest = request.toHTTPRequest();
+
+				// convert the Nimbus HTTPRequest into an Apache HttpClient HttpRequest
+				HttpRequestBase apacheHttpRequest;
+				switch (httpRequest.getMethod()) {
+					case GET:
+						String url = Misc.concatStrings(httpRequest.getURL().toExternalForm(), "?", httpRequest.getQuery());
+						apacheHttpRequest = new HttpGet(url);
+						break;
+					case POST:
+						apacheHttpRequest = new HttpPost(httpRequest.getURL().toExternalForm());
+						((HttpPost)apacheHttpRequest).setEntity(new StringEntity(httpRequest.getQuery()));
+						break;
+					default:
+						throw new IllegalStateException("Illegal Method, must be GET or POST");
+				}
+				httpRequest.getHeaderMap().forEach((k,l) -> l.forEach(v -> apacheHttpRequest.addHeader(k, v)));
```

Comment:
Misschien dat je deze lange methode op kan delen in stukjes, zoals het opstellen van het request (en/of ook het verwerken er van).
Er zijn nu 4 try blocken.